applications/dashboard/controllers/api/AddonsApiController.php

```diff
@@ -0,0 +1,210 @@
+            return strcasecmp($a['name'], $b['name']);
+        });
+        $addons = $out->validate($addons);
+
+        return $addons;
+    }
+
+    /**
+     * Get a single addon.
+     *
+     * @param string $addonID The addon ID to get.
+     * @param array $query The query string.
+     * @return array Returns an addon info row.
+     * @throws \Garden\Web\Exception\NotFoundException Throws an exception when the addon is not found.
+     */
+    public function get($addonID, array $query) {
+        $this->permission('Garden.Settings.Manage');
+
+        $id = $this->schema([
+            'addonID:s' => 'The ID of the addon.'
+        ], 'in')->setDescription('Get an addon.');
+
+        $in = $this->schema([
+            'themeType:s?' => [
+                'description' => 'Which theme to show the enabled status for.',
+                'enum' => ['desktop', 'mobile'],
+                'default' => 'desktop'
+            ]
+        ], 'in');
+        $out = $this->schema($this->fullSchema(), 'out');
+        $query = $in->validate($query);
+
+        $addons = $this->addonModel->getWhere(['addonID' => $addonID]);
+        if (empty($addons)) {
+            throw new \Garden\Web\Exception\NotFoundException("Addon");
+        }
+
+        $row = $out->validate($this->filterOutput(reset($addons)));
+        return $row;
+    }
+
+    /**
+     * Enable or disable an addon.
+     *
+     * @param string $addonID The addon ID.
+     * @param array $body The request body.
+     * @return array Returns an array of affected addons.
+     * @throws \Garden\Web\Exception\NotFoundException Throws an exception when the addon is not found.
+     */
+    public function patch($addonID, array $body) {
+        $this->permission('Garden.Settings.Manage');
+
+        $idSchema = $this->schema([
+            'addonID:s' => 'The ID of the addon.'
+        ], 'in');
+        $in = $this->schema([
+            'enabled:b' => 'Enable or disable the addon.',
+            'themeType:s?' => [
+                'description' => 'Which theme type to set.',
+                'enum' => ['desktop', 'mobile'],
+                'default' => 'desktop'
+            ]
+        ], 'in')->setDescription('Enable or disable an addon.');
+        $out = $this->schema([':a' => $this->fullSchema()], 'out');
+
+        $body = $in->validate($body);
+
+        $addons = $this->addonModel->getWhere(['addonID' => $addonID]);
+        if (empty($addons)) {
+            throw new \Garden\Web\Exception\NotFoundException("Addon");
+        }
+        $addon = reset($addons);
+
+        $options = [];
+        if (!empty($body['themeType'])) {
+            $options['themeType'] = $body['themeType'];
+        }
+        if ($body['enabled']) {
+            $r = $this->addonModel->enable($addon, $options);
+        } else {
+            $this->addonModel->disable($addon, $options);
+            $r = [$addon];
+        }
+
+        $r = array_map([$this, 'filterOutput'], $r);
```

Comment:
This fails on addons with dependencies (e.g. Ideation). The result of `AddonModel::enable` for these types of addons has the first element being an instance of `Vanilla\Addon` and subsequent elements (newly-enabled requirements) being associative arrays. The type declaration for the first parameter of `AddonsApiController::filterOutput` is `Vanilla\Addon`.

Reply:
I fixed this in the PR, but need to hold off on unit tests. There is another PR coming that has a place for this kind of unit test.